### ISSUE TITLE
Optional logging of all events to a file

### ISSUE THREAD

**Issue by magnusbaeck** [boivie/lovebeat#2]
I think it would be useful to have a way of dumping all events flowing through the event bus to a file in machine-readable JSON. This would make it trivial to ingest the events into e.g. Logstash, NXLog, or Fluentd for further dispatching to e.g. brokers like Redis, RabbitMQ, or Kafka, or just storing the events somewhere—and without constant feature creep in Lovebeat to support more and more outputs. Example of what such a file could look like:

    {"timestamp":"2015-05-31T15:03:00Z","type":"ViewStateChangedEvent","event":<serialized ViewStateChangedEvent>}

If this seems reasonable I can provide the patch.

**Comment by boivie** [comment]
I totally agree - and have been wanting something like this to feed back into another system. Please go ahead - and we can also have a discussion offline if you want to.

**Comment by magnusbaeck** [comment]
Mostly done. The log entries look like this:

    {"ServiceStateChangedEvent":{"Service":{"Name":"invoice.mailer","LastValue":-1,"LastBeat":1436812717893,"BeatHistory":null,"WarningTimeout":-1,"ErrorTimeout":-1,"State":"ok"},"Previous":"paused","Current":"ok"},"Timestamp":"2015-07-13T18:38:37.893697584Z","Type":"ServiceStateChangedEvent"}

An event `ev` is marshalled with

    t := reflect.TypeOf(ev).Name()
    jev := map[string]interface{}{
            "Timestamp": time.Now().UTC(),
            "Type":      t,
            t:           &ev,
    }

i.e. the Type property indicates the name of the property holding the actual event struct. Seems reasonable?

I'm a bit annoyed about having to pass all the event types to eventbus.RegisterHandler when registering the event handler for the JSON logging. I'm inclined to figure out a nice way to have it send _all_ events to handlers whose argument is interface{}.

I'll use the file rotation aware file writer from https://github.com/mipearson/rfw.

**Comment by boivie** [comment]
I didn't know about rfw - it looks like the perfect tool for the job.

When it comes to the output json, I think we should treat it as an external API and re-encode every specific type. The messages that goes on the event bus are highly internal and we might choose to add fields on those structures that should not be persisted in the files. Yes, it's annoying.

At the same time, I would prefer to use lowercase field names for the written events, like:

```
{
    "ts": "2015-07-13T18:38:37.893697584Z",
    "type": "service_state_changed",
    "service_state_changed": {
        "service": {
            "name": "invoice.mailer",
            "last_beat": 1436812717893,
            "beat_history": null,
            "warning_timeout": -1,
            "error_timeout": -1,
            "state": "ok"
        },
        "previous": "paused",
        "current": "ok"
    }
}
```

Does this make sense?

**Comment by magnusbaeck** [comment]
Sure, makes total sense. It might get hairy if we want to restructure the events completely when we marshal them, but if it's just a matter of renaming fields and omitting a few boring ones then it's trivial. If we view this as an external API there should definitely be tests that verify exactly what JSON we get so that we don't leak data.

Do you want the log itself to contain pretty-printed JSON or was that just for readability purposes here?

**Comment by boivie** [comment]
Absolutely - tests would be good. We have some code coverage setup at https://coveralls.io/github/boivie/lovebeat and the numbers are not that great yet ;)

The pretty-printing was for readability purposes. I guess it's best to have them on a single line since the file can be split.

**Comment by magnusbaeck** [comment]
Closed with the merge of #6.
